### PR TITLE
Convert remaining legacy modules in RECO to support multi-threading

### PR DESCRIPTION
**Pull request by cvuosalo**
Converted the last remaining legacy reco modules that were taking some amount of time, as reported by igprof.

This PR is a copy of #10806, with merge conflicts fixed.

### CODE REVIEW

**Comment by cvuosalo** [comment]
@cmsbuild please test

File: PhysicsTools/PatAlgos/plugins/PATElectronProducer.h [hunk 1]
```diff
@@ -166,70 +166,63 @@ namespace pat {
...
-    if (depconf.exists("pfPhotons")) {
-      labels.push_back(std::make_pair(pat::PfGammaIso, depconf.getParameter<edm::InputTag>("pfPhotons")));
-    }
-    if (depconf.exists("user")) {
-      std::vector<edm::InputTag> userdeps = depconf.getParameter<std::vector<edm::InputTag> >("user");
-      std::vector<edm::InputTag>::const_iterator it = userdeps.begin(), ed = userdeps.end();
-      int key = pat::IsolationKeys::UserBaseIso;
-      for ( ; it != ed; ++it, ++key) {
-       labels.push_back(std::make_pair(pat::IsolationKeys(key), *it));
+																							 const char* psetName,
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATElectronProducer.h]: this became quite unreadable ( ``` const char* psetName, ``` and other arguments ended up on column ~200)

**Comment by slava77** [comment]
@cvuosalo I suggest to update the indentation in PATElectronProducer.h 
Other changes appear the same as they were in #10806 or the equivalent code from the head/release was taken instead of #10806 (e.g. folding PATMuonProducer.h and PATPhotonProducer.h parts into the namespace scope; these were mostly cosmetic and I don't think that they have to be put back in).

**Comment by slava77** [comment]
@cvuosalo 
I see a recent merge commit.
Which release is used?

**Comment by cvuosalo** [comment]
@slava: The IB used was CMSSW_7_6_X_2015-09-03-1100, the same as before.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cvuosalo I checked the last updates and they seem fine now 

BTW, please make PRs only from some descriptive topic branches (not from "merge-attempt")

**Comment by cvuosalo** [comment]
+1

For #11140 7366661d39e57fb53b3ae86d5db46ceedb3e6c6e

Convert remaining legacy modules in RECO to support multi-threading. There should be no change in monitored quantities. This PR is derived from #10806 and is intended to fix its merge conflicts.

Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11140/7988/summary.html) against baseline CMSSW_7_6_X_2015-09-09-1100 show no significant differences, as expected. See #10806 for information about the performance of this PR.

**Comment by lgray** [comment]
@slava77 @cvuosalo I will start from this branch and construct new backports.

@davidlange6 I'll close the old ones.

**Pull request merged**